lib/lhm/progress.rb

```diff
@@ -0,0 +1,195 @@
+      @completion_percentage = ((@bytes_copied.to_f / @total_bytes) * 100.00).round(2)
+    
+      @speedometer << @bytes_copied
+      @copy_speed = @speedometer.speed
+
+      current_time = Time.now
+      if (current_time - @start_time > (5 * 60)) || @options.include?(:verbose_logging)
+        @printer.notify(serialized_progress)
+        @start_time = current_time
+      end
+    end
+
+    def update_state(state)
+      populate_table_stats if state == Lhm::STATE_COPYING
+      @completion_percentage = 100 if state == Lhm::STATE_COPYING_DONE
+
+      @state = state
+      @printer.notify(serialized_progress)
+    end
+
+    def serialized_progress
+      JSON.pretty_generate({
+        "state" => @state,
+        "version" => Lhm::VERSION,
+        "source_table" => @origin.name,
+        "destination_table" => @origin.destination_name,
+        "throttler" => @throttler,
+        "table_switcher" => @table_switcher,
+        "avg_row_length" => "#{@avg_row_length} bytes",
+        "estimated_bytes_copied" => @bytes_copied,
+        "estimated_total_bytes" => @total_bytes,
+        "completion_percentage" => "#{@completion_percentage}% complete",
+        "estimated_copy_speed" => "#{@copy_speed} bytes/sec",
+        "min_pk_key" => @min_pk_key,
+        "max_pk_key" => @max_pk_key,
+        "rows_written" => @rows_written,
+      })
+    end
+
+    private
+
+    def populate_table_stats
+      table_stats = get_table_stats_from_information_schema
+      @avg_row_length = table_stats[0]
+      @total_bytes = table_stats[1]
+    end
+
+    def get_table_stats_from_information_schema
+      query = %W{
+        SELECT
+          avg_row_length, (data_length + index_length)
+        FROM information_schema.tables
+        WHERE table_schema = '#{ActiveRecord::Base.connection.current_database}'
+        AND table_name = '#{@origin.name}'
+      }
+      @retry_helper.with_retries do |retriable_connection|
+        retriable_connection.select_rows(query.join(' ')).first
+      end
+    end
+
+    # Speedometer is used to calculate the speed at which LHM is copying rows from source to the destination table. 
+    # Taken from the PR authored by @shuhaowu from here - https://github.com/Shopify/lhm/pull/83. 
+    class Speedometer
+      attr_reader :log
+
+      def self.linregress(x, y)
+        raise ArgumentError, "x and y not the same length" if x.length != y.length
+
+        n = x.length
+        xsum = 0.0
+        ysum = 0.0
+        xxsum = 0.0
+        yysum = 0.0
+        xysum = 0.0
+
+        n.times do |i|
+          xsum += x[i]
+          ysum += y[i]
+          xxsum += x[i] ** 2
+          yysum += y[i] ** 2
+          xysum += x[i] * y[i]
+        end
+
+        denom = (n * xxsum - xsum ** 2)
+        if denom == 0
+          return [0, 0, true]
+        end
+
+        slope = (n * xysum - xsum * ysum) / denom
+        intercept = ysum / n - slope * xsum / n
+
+        [slope, intercept, false]
```

Comment:
Nit: Adding true or false at the end is very "Go-like". Since you're using defensive programming and returning nil on `linregress` being singular, maybe returning an exception would be more consistent with Ruby and provide the same result.